### DISCUSSION TITLE
Diagonal heatmap

### DISCUSSION THREAD

**Discussion by ekaterinaesenkova** [jokergoo/ComplexHeatmap#773] [Q&A]
Good day!

Is there a possibility for a diagonal heatmap ( to have only upper/lower triangular of it plotted)?

Thanks in advance!

**Comment by jokergoo** [comment]
Check this example code:

```r
m = matrix(rnorm(100), 10)
cm = cor(m)

od =  hclust(dist(cm))$order

cm = cm[od, od]

Heatmap(cm, rect_gp = gpar(type = "none"), cluster_rows = FALSE, cluster_columns = FALSE,
	cell_fun = function(j, i, x, y, w, h, fill) {
		if(i >= j) {
			grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
		}
	})
```

One thing needs to be noted: Here we compare the row index and column index in `cell_fun`, this requires the matrix for `Heatmap()` (which is `cm`) is already reordered by clustering before sending to heatmap.

<img src="https://user-images.githubusercontent.com/449218/126462627-ed422cb3-b69e-448d-b0c9-ba6e191127ed.png" width="600" />

**Reply by ekaterinaesenkova** [reply]
Thank you! One more question: is there a possibility to keep dendrograms, previously created by hclust (as cluster_rows = FALSE, cluster_columns = FALSE)? 

Thanks in advance!

**Comment by jokergoo** [comment]
Then I think you can directly compare `x` and `y` which are the physical locations in the heatmap body. In this case, you don't need to pre-order the matrix and instead, you can apply clustering to the matrix:

```r
m = matrix(rnorm(100), 10)
cm = cor(m)

Heatmap(cm, rect_gp = gpar(type = "none"), column_dend_side = "bottom",
	cell_fun = function(j, i, x, y, w, h, fill) {
		if(as.numeric(x) <= 1 - as.numeric(y)) {
			grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
		}
	})
```
![image](https://user-images.githubusercontent.com/449218/126624732-531c4fa0-146b-4fa5-bf8f-e8b45a0e29dd.png)

**Reply by jokergoo** [reply]
Here we compare two floating numbers. As you can see, there is one white square on the diagonal. I suggest to add a tiny positive offset to the right side of the comparison:

```r
Heatmap(cm, rect_gp = gpar(type = "none"), column_dend_side = "bottom",
	cell_fun = function(j, i, x, y, w, h, fill) {
		if(as.numeric(x) <= 1 - as.numeric(y) + 1e-6) {
			grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
		}
	})
```

![image](https://user-images.githubusercontent.com/449218/126625997-56440598-8838-42c2-be3b-37eddd0b9676.png)

**Reply by ekaterinaesenkova** [reply]
Huge thanks! Another suggestion is to have the function to combine two diagonal heatmaps together (as upper and lower triangular of the heatmap).

**Reply by jokergoo** [reply]
You can check the last plot from this post (https://jokergoo.github.io/2021/05/11/multiple-color-themes-in-a-single-heatmap/). Directly writing a general function for this might be difficult, but with a little bit effort with `cell_fun`, I think it is possible.

**Reply by ekaterinaesenkova** [reply]
Thanks for sharing! 
Maybe useful for others: I created two opposite diagonal heatmaps (with sign > change in the code you provided above) and then I put them in ht_list = h1 + h2 and draw with negative unit distance until they collapse:
draw(ht_list,annotation_legend_side = "top", heatmap_legend_side = "top",ht_gap = unit(-42, "cm"))

Not the most elegant way to do it, but it worked for me :)

**Reply by jokergoo** [reply]
Ha, good idea! 👍

**Reply by jokergoo** [reply]
I have put this method in my blog post: https://jokergoo.github.io/2021/07/22/make-triangle-heatmap/. Thanks again!

**Reply by ekaterinaesenkova** [reply]
Thank you!  Another problem I encountered - when I create ht_list to combine the diagonal matrices, it takes just row annotation from my h1 and duplicates it on my h2 side as well, so my original row annotation for h2 is overwritten. (see the image: the right side should be the same as upper annotation).  Is there a way to solve this?  Thanks in advance!
![Screenshot 2021-07-25 at 15 17 25](https://user-images.githubusercontent.com/83301670/126900484-21eac6db-9da0-435e-9710-8f0991e0b0b2.png)

**Reply by jokergoo** [reply]
Can you also attach the code? The following is what I have tried and it works fine.

```r
library(ComplexHeatmap)
m = cor(matrix(rnorm(100), 10))

library(circlize)
col1 = colorRamp2(c(-1, 0, 1), c("green", "black", "red"))
col2 = colorRamp2(c(-1, 0, 1), c("purple", "white", "orange"))

od =  hclust(dist(m))$order
m2 = m[od, od]

ht1 = Heatmap(m2, rect_gp = gpar(type = "none"), col = col1,
    cluster_rows = FALSE, cluster_columns = FALSE,
    bottom_annotation = HeatmapAnnotation(foo1 = 1:10),
    left_annotation = rowAnnotation(bar1 = 1:10),
    cell_fun = function(j, i, x, y, w, h, fill) {
        if(i >= j) {
            grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
        }
    })

ht2 = Heatmap(m2, rect_gp = gpar(type = "none"), col = col2,
    cluster_rows = FALSE, cluster_columns = FALSE,
    top_annotation = HeatmapAnnotation(foo2 = rep(c("a", "b"), each = 5)),
    right_annotation = rowAnnotation(bar2 = rep(c("a", "b"), each = 5)),
    cell_fun = function(j, i, x, y, w, h, fill) {
        if(i <= j) {
            grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
        }
    })

draw(ht1 + ht2, ht_gap = unit(-70, "mm"))
```

![image](https://user-images.githubusercontent.com/449218/126912029-baf3ecd8-d4e2-45ab-be2d-ba3dd4d19f85.png)

**Reply by ekaterinaesenkova** [reply]
Thanks for the response! Below is my code for annotation for my genomic and metabolic distance matrices.  I noticed that if I change the order of ht_list= h1 + h2 to ht_list= h2 + h1, I get annotation of rows for h2. Thanks in advance!

```
#genomic 
column_ha = HeatmapAnnotation(Phylum = as.factor(phylum),
col = list(Phylum = c(Actinobacteria ="#0F02C2", Bacteroidetes = "#3C9953", Firmicutes = "#FFED7D", Proteobacteria = "#C22741")), show_legend = c("Phylum" = FALSE), border = TRUE)

row_ha = rowAnnotation(Phylum = as.factor(phylum), Class = as.factor(class),Gsize=anno_points(as.numeric(Gsize)), col = list(Phylum = c( Actinobacteria ="#0F02C2", Bacteroidetes = "#3C9953", 
Firmicutes = "#FFED7D", Proteobacteria = "#C22741"), Class = c(Actinomycetia="#0F02C2", Bacilli="#1CEFFF", Bacteroidia="#3C9953", Clostridia="#FFED7D", Coriobacteriia="#D20AFF",Erysipelotrichia="#C22741", Gammaproteobacteria="#FFC2FA")), show_legend = c("Phylum" = FALSE, "Class" = FALSE),  border = TRUE)

col_fun = colorRamp2(c(0, 0.5, 1), c("white", "grey","black"))


h1=Heatmap(as.matrix(gen_dist), rect_gp = gpar(type = "none"), column_dend_side = "bottom", name = "Genomic similarity", bottom_annotation = column_ha, left_annotation = row_ha,
           heatmap_legend_param = list(direction = "horizontal",side="top"),
           col=col_fun,
           cell_fun = function(j, i, x, y, w, h, fill) {
             if(as.numeric(x) <= 1 - as.numeric(y) + 1e-6) {
               grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
             }
           })

#metabolic

column_ha = HeatmapAnnotation(Phylum = as.factor(phylum), 
col = list(Phylum = c(Actinobacteria ="#0F02C2", Bacteroidetes = "#3C9953", Firmicutes = "#FFED7D", Proteobacteria = "#C22741")),border = TRUE)
row_ha= rowAnnotation(Phylum = as.factor(phylum), Class = as.factor(class),  
col = list(Phylum2 = c(Actinobacteria ="#0F02C2", Bacteroidetes = "#3C9953", Firmicutes = "#FFED7D", Proteobacteria = "#C22741"), Class2 = c(Actinomycetia="#0F02C2", Bacilli="#1CEFFF", Bacteroidia = "#3C9953", Clostridia="#FFED7D", Coriobacteriia="#D20AFF", Erysipelotrichia="#C22741",Gammaproteobacteria="#FFC2FA")),show_legend = c("Phylum" = FALSE, "Class" = TRUE),border = TRUE)

h2=Heatmap(as.matrix(met_dist), rect_gp = gpar(type = "none"), column_dend_side = "top",row_dend_side = "right", 
           name = "Metabolic distance", top_annotation = column_ha, right_annotation = row_ha,
           col=col_fun, heatmap_legend_param = list(direction = "horizontal",side="right"),
           cell_fun = function(j, i, x, y, w, h, fill) {
             if(as.numeric(x) >= 1 - as.numeric(y) + 1e-6) {
               grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
             }
           })


ht_list = h1+h2
draw(ht_list,annotation_legend_side = "bottom", heatmap_legend_side = "bottom",ht_gap = unit(-47, "cm"))

```

**Reply by jokergoo** [reply]
If you change to `h2 + h1`, then you should also change the settings in the two heatmaps, i.e., change `top_annotation` to `bottom_annotation` in h2, and `right_annotation` to `left_annotation` in h2. The direction of the comparisons in two `cell_fun` should also be reversed.

**Reply by ekaterinaesenkova** [reply]
thank you!

**Comment by Sbzahedi** [comment]
Good day! Is it possible to split in diagonal while keeping cluster_rows and cluster_columns as true?

**Reply by Sbzahedi** [reply]
Here is my code so far: 

corr_mx <- round(cor(log10(map + 1)), 2)
coul <- rev(colorRampPalette(brewer.pal(8, "Spectral"))(25))

row_dend <- hclust(dist(corr_mx))
column_dend <- hclust(dist(t(corr_mx)))

row_cut <- cutree(row_dend, h = 0.5*max(row_dend$height))
column_cut <- cutree(column_dend, h = 0.5*max(column_dend$height))

hm <- Heatmap(as.matrix(corr_mx), col = coul,
                           heatmap_legend_param = list(title = "Legend", direction = "horizontal"),
                           cluster_rows = TRUE, cluster_columns = TRUE,
                           row_split = row_cut, column_split = column_cut,
                           row_names_gp = gpar(fontsize = 2), column_names_gp = gpar(fontsize = 2),
                           column_title = "HEATMAP", 
                           column_title_gp = gpar(fontsize = 15, fontface = "bold"), 
                           column_title_side = "bottom",
                           show_column_dend = FALSE, show_row_names = FALSE, row_dend_width = unit(1, "cm"),
                           row_gap = unit(0.5, "mm"), column_gap = unit(0.5, "mm"),
                           layer_fun = function(j, i, x, y, width, height, fill, slice_r, slice_c) {
                             v = pindex(corr_mx, i, j)
                             if (slice_r == slice_c) {
                               grid.rect(gp = gpar(lwd = 1.5, fill = "transparent"))}}
)
draw(hm, heatmap_legend_side = "bottom")

**Comment by sotuamax** [comment]
Greetings!

I try to draw a very large matrix in triangle form using Heatmap cell_fun. However, it takes long time to draw it. I checked the document, and layer_fun is a vectorized version of cell_fun and may runs faster. Is that possible to provide a example code here to draw triangle heatmap using layer_fun? 

Many thanks. 

Best,

**Reply by jokergoo** [reply]
Can you also provide some example data and code?

**Reply by sotuamax** [reply]
Hello, 

I attached a subset data here. 
[counts.txt](https://github.com/user-attachments/files/16997078/counts.txt)

The code I used to draw a triangle heatmap in a faster way (cell_fun works but is very slow): 

mat <- read.table("counts.txt", sep = "\t", header = F) |> as.matrix()
col_fun <- colorRamp2(c(1, 3, 5, 8, 10), c("white", "#FBB91F", "#F17020", "#C43C4E", "#9C2964FF"))
Heatmap(log2(mat+1), 
        name = "ht1", 
        rect_gp = gpar(type = "none"), 
        cluster_rows = F, 
        cluster_columns = F, 
        show_row_dend = F, 
        show_column_dend = F, 
        show_column_names = F, 
        show_row_names = F, 
        col = col_fun, 
        border = T, 
        show_heatmap_legend = F, 
        width = unit(0.8, "snpc"), 
        height = unit(0.8, "snpc"), 
        layer_fun = function(j, i, x, y, width, height, fill) {
          v = pindex(mat, i, j)
          if (v[i, ] < v[, j]) {
          grid.rect(gp = gpar(lwd = 1.5, fill = fill, col = fill))}}
)


I also try to rotate the triangle heatmap and save it to a pdf file (like below). 
![Screenshot 2024-09-13 at 1 09 10 PM](https://github.com/user-attachments/assets/fec8a5f9-19fe-4402-8036-941faf34fe1c)

However, in reference to https://github.com/jokergoo/ComplexHeatmap/issues/186, it doesn't show well. Is there any better way to 
do it? 

The plot is for a HiC related, and I called topological associated domain (TAD) from other source, and annotated the region. I can mark the TAD use square like below. Is there any way to mark it triangle of TAD? 
![Screenshot 2024-09-13 at 1 15 43 PM](https://github.com/user-attachments/assets/f614f6e8-5c20-4747-b13a-23d0eca99594)


Thank you so much!

Best,

**Comment by jokergoo** [comment]
I will use the `mat[100, 100]` for demonstration.

```r
Heatmap(log2(mat[1:100, 1:100]+1),
	name = "ht1",
	rect_gp = gpar(type = "none"),
	cluster_rows = F,
	cluster_columns = F,
	show_row_dend = F,
	show_column_dend = F,
	show_column_names = F,
	show_row_names = F,
	col = col_fun,
	border = T,
	show_heatmap_legend = F,
	width = unit(0.8, "snpc"),
	height = unit(0.8, "snpc"),
	layer_fun = function(j, i, x, y, width, height, fill) {
		l = i >= j
		grid.rect(x[l], y[l], width[l], height[l], gp = gpar(lwd = 1.5, fill = fill[l], col = fill[l]))
	}
)
```

As in `layer_fun`, all arguments are vectors, to only draw the triangle, e.g. the lower triangle, you just need to use the values where `i >= j`.

<img width="783" alt="image" src="https://github.com/user-attachments/assets/70558cf6-389d-44d3-9324-9f950d2d13db">

To rotate the plot by 45 degrees, you can simply push a global viewport with 45 degrees rotation:

```r
ht = Heatmap(log2(mat[1:100, 1:100]+1),
	name = "ht1",
	rect_gp = gpar(type = "none"),
	cluster_rows = F,
	cluster_columns = F,
	show_row_dend = F,
	show_column_dend = F,
	show_column_names = F,
	show_row_names = F,
	col = col_fun,
	border = T,
	show_heatmap_legend = F,
	width = unit(0.8, "snpc"),
	height = unit(0.8, "snpc"),
	layer_fun = function(j, i, x, y, width, height, fill) {
		l = i >= j
		grid.rect(x[l], y[l], width[l], height[l], gp = gpar(lwd = 1.5, fill = fill[l], col = fill[l]))
	}
)

grid.newpage()
pushViewport(viewport(angle = 45, width = 1/sqrt(2), height = 1/sqrt(2)))
draw(ht, newpage = FALSE)
popViewport()
```

<img width="888" alt="image" src="https://github.com/user-attachments/assets/7c920f35-d86d-4d3b-a941-1af3803143c5">

**Comment by sotuamax** [comment]
Thank you so much for your prompt response. 

It works! 

Best,
Meiyuan
